sw/host/opentitanlib/src/util/openocd.rs

```diff
@@ -67,22 +69,32 @@ impl OpenOcdServer {
     }
 
     /// Wait until we see a particular message on the output.
-    fn wait_until_message(stdout: &mut impl Read, msg: &str, timeout: Duration) -> Result<()> {
-        let mut s: Vec<u8> = Vec::new();
+    fn wait_until_regex_match<'a>(
+        stderr: &mut impl Read,
+        regex: &Regex,
+        timeout: Duration,
+        s: &'a mut Vec<u8>,
+    ) -> Result<regex::Captures<'a>> {
         let start = Instant::now();
         loop {
             let mut byte = 0u8;
             // NOTE the read could block indefinitely, a proper solution would involved spawning
             // a thread or using async.
-            let n = stdout.read(std::slice::from_mut(&mut byte))?;
+            let n = stderr.read(std::slice::from_mut(&mut byte))?;
             if n != 1 {
                 bail!("OpenOCD stopped before being ready?");
             }
             if byte == b'\n' {
-                let string = String::from_utf8_lossy(&s[..]);
+                let string = std::str::from_utf8(&s[..])?;
                 log::info!(target: concat!(module_path!(), "::stderr"), "{}", string);
-                if string == msg {
-                    return Ok(());
+                if let Some(_captures) = regex.captures(string) {
+                    // I would have wanted to return Ok(_captures) here, but doing so causes the
+                    // Rust compiler to complain that returning a value, which needs to immutably
+                    // borrow for 'a lifetime somehow conflicts with the mutable borrowing in
+                    // other branches further down this loop.  Moving the return statement outside
+                    // the loop eliminates the error, but requires duplicate invocation of the
+                    // regex method.
+                    break;
```

Comment:
I've reduced this to the following minimal example:

```rust
fn foo<'a>(vec: &'a mut Vec<u8>) -> &'a u8 {
    if let Some(value) = vec.get(0) {
        return value;
    }
    vec.push(0);
    return &vec[0];
}
```

This fails to compile with normal `rustc` flags, but succeeds with `RUSTFLAGS="-Zpolonius"`. That indicates the error is probably a result of borrow checker limitations and not a fundamental issue with the code.

Therefore, I believe you are correct to invoke the regex twice. I suggest using `regex.is_match(string)` in the loop to detect whether the regex matches, to avoid running the full capture engine twice.